src/Http/Request.php

```diff
@@ -0,0 +1,332 @@
+      static::METHOD_CONNECT,
+    ));
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function isValidMethod($method, $strict = TRUE) {
+    $method = $strict ? $method : strtolower($method);
+    return static::isReadMethod($method, $strict) || static::isWriteMethod($method, $strict);
+  }
+
+  /**
+   * Parses the body string.
+   *
+   * @return array
+   */
+  public function getParsedBody() {
+    if ($this->parsedBody) {
+      return $this->parsedBody;
+    }
+    // Find out the body format and parse it into the \ArrayObject.
+    $this->parsedBody = static::parseBody($this->method);
+    return $this->parsedBody;
+  }
+
+  /**
+   * Parses the body.
+   *
+   * @param string $method
+   *   The HTTP method.
+   *
+   * @return array
+   *   The parsed body.
+   */
+  protected static function parseBody($method) {
+    $body = NULL;
+    if ($method == static::METHOD_GET) {
+      return $_GET;
+    }
+    elseif ($method == static::METHOD_POST) {
+      return $_POST;
+    }
+
+    if ($query_string = file_get_contents('php://input')) {
+      // When trying to POST using curl on simpleTest it doesn't reach
+      // $_POST, so we try to re-grab it here.
+      // Also, sometimes the client might send the input still encoded.
+      if ($decoded_json = drupal_json_decode($query_string)) {
+        return $decoded_json;
+      }
+      else {
+        parse_str($query_string, $body);
+        return $body;
+      }
+    }
+
+    return NULL;
+  }
+
+  /**
+   * Parses the header names and values from globals.
+   *
+   * @return HttpHeaderBag
+   *   The headers.
+   */
+  protected static function parseHeadersFromGlobals() {
+    $bag = new HttpHeaderBag();
+    $headers = array();
+    if (function_exists('apache_request_headers')) {
+      $headers = apache_request_headers();
+    }
+    else {
+      foreach ($_SERVER as $key => $value) {
+        if (strpos($key, 'HTTP_') === 0) {
+          // Generate the plausible header name based on the $name.
+          // Converts 'HTTP_X_FORWARDED_FOR' to 'X-Forwarded-For'
+          $name = substr($key, 5);
+          $parts = explode('_', $name);
+          $parts = array_map('strtolower', $parts);
+          $parts = array_map('ucfirst', $parts);
+          $name = implode('-', $parts);
+          $headers[$name] = $value;
+        }
+      }
+    }
+    // Iterate over the headers and bag them.
+    foreach ($headers as $name => $value) {
+      $bag->bag(HttpHeader::create($name, $value));
+    }
+    return $bag;
+  }
+
+  /**
+   * Gets the fully qualified URL with the query params.
+   *
+   * @return string
+   *   The URL.
+   */
+  public function href() {
+    return url($this->path, array(
+      'absolute' => TRUE,
+      'query' => $this->query,
+    ));
+  }
+
+  /**
+   * Get the credentials based on the $_SERVER variables.
+   *
+   * @return array
+   *   A numeric array with the username and password.
+   */
+  protected static function getCredentials() {
```

Comment:
I think understand why you moved it to ``Request``, since it's dealing with ``$_SERVER``, but it feels it shouldn't be here